modules/web-console/src/main/java/org/apache/ignite/console/routes/AgentDownloadRouter.java

```diff
@@ -0,0 +1,149 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.console.routes;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.InputStream;
+import java.nio.file.Paths;
+import java.util.Enumeration;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+import java.util.zip.ZipOutputStream;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.http.HttpServerResponse;
+import io.vertx.ext.auth.User;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+import org.apache.ignite.Ignite;
+
+/**
+ * Router to handle REST API to download Web Agent.
+ */
+public class AgentDownloadRouter extends AbstractRouter {
+    /** Buffer size of 30Mb to handle Web Agent ZIP file manipulations. */
+    private static final int BUFFER_SZ = 30 * 1024 * 1024;
+
+    /** */
+    private File pathToWebAgentZip;
+
+    /** */
+    private final String file;
+
+    /**
+     * @param ignite Ignite.
+     */
+    public AgentDownloadRouter(Ignite ignite, String folder, String file) {
+        super(ignite);
+
+        this.file = file;
+        this.pathToWebAgentZip = Paths.get(folder, file + ".zip").toFile();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void initializeCaches() {
+        // No-op so far.
+    }
+
+    /** {@inheritDoc} */
+    @Override public void install(Router router) {
+        router.get("/api/v1/downloads/agent").handler(this::load);
+    }
+
+    /**
+     * @param ctx Context.
+     */
+    private void load(RoutingContext ctx) {
+        User user = checkUser(ctx);
+
+        if (user != null) {
+            try {
+                ZipFile zip = new ZipFile(pathToWebAgentZip);
+
+                ByteArrayOutputStream baos = new ByteArrayOutputStream(BUFFER_SZ);
+
+                ZipOutputStream zos = new ZipOutputStream(baos);
+
+                Enumeration<? extends ZipEntry> entries = zip.entries();
+
+                byte[] buf = new byte[BUFFER_SZ];
+
+                while (entries.hasMoreElements()) {
+                    ZipEntry e = entries.nextElement();
+
+                    zos.putNextEntry(e);
+
+                    if (!e.isDirectory()) {
+                        int bytesRead;
+
+                        InputStream is = zip.getInputStream(e);
+
+                        while ((bytesRead = is.read(buf)) != -1)
+                            zos.write(buf, 0, bytesRead);
+                    }
+
+                    zos.closeEntry();
+                }
+
+                // Append default.properties to ZIP.
+                ZipEntry e = new ZipEntry(file + "/default.properties");
+                zos.putNextEntry(e);
+
+                String content = String.join("\n",
+                    "tokens=" + user.principal().getString("token", "MY_TOKEN"),
+                    "server-uri=" + ctx.request().host(),
```

Comment:
This may incorrect work under proxy. Please look at `modules/web-console/backend/middlewares/host.js` logic

Reply:
Fixed.